### DISCUSSION TITLE
Can guidance generate list of jsons?

### DISCUSSION THREAD

**Discussion by CSWellesSun** [guidance-ai/guidance#402] [Q&A] [answered]
I want to generate the following format, that is, list of jsons:
[
    {"name": "Alice", "age": 1},
    {"name": "Bob", "age": 2},
]
The number of the objects in the list is random depending on the output of LLM. So can guidance support such format?

**Comment by motaatmo** [comment]
Hi,

I have been trying to achieve something similar. Basically I think it would be nice to have a direct way to parse JSONs using a Guidance-grammar. I tried to write a very basic version to match my needs (i.e. short of an implementation of the JSON schema, which would certainly be the best way to do it). I failed. Using GPT3.5, I went the ugly way to simply ask the model for a string and to load it with json.loads(). As the results are most of the time well-formed JSON strings, this works for me, although I would, of course, prefer to really guide the model (and not just hope that it spits out a string that json.loads can process).

**Reply by CSWellesSun** [reply]
I achieved the list of json with the former version. But recently, Guidance has a huge update and changes all the syntax, so I haven't figured out how to achieve the effect in the current version. If I have any idea, I will comment later.

**Comment by marcotcr** [comment, marked as answer]
I will push something general for json at some point, but here is a version for your simple use case:
```python
from guidance import select, zero_or_more, any_char, one_or_more, commit_point, regex

@guidance(stateless=True)
def simple_json(lm):
    lm += ('{"name": ' + one_or_more(any_char()) + commit_point('"') +
            ', "age": ' + regex('\d+') + commit_point('}') + ',\n')
    return lm

@guidance(stateless=True)
def simple_list_of_jsons(lm):
    lm += '[\n' + one_or_more(simple_json()) + ']'
    return lm
```
The `commit_point`s are just ways of stopping `one_or_more(any_char())` once we hit `"`, or `\d+` ondce we hit `'}'`.
Notice that this does not constrain the LM, i.e. it can generate jsons forever. Here is mistral-7b though, it seems to know when to stop if the instruction is clear
```python
mistral + 'Please generate a list of 5 jsons with name and age:\n' + simple_list_of_jsons()
```
> Please generate a list of 5 jsons with name and age:
[
{"name": "John", "age": 20},
{"name": "Jane", "age": 25},
{"name": "Bob", "age": 30},
{"name": "Alice", "age": 35},
{"name": "Charlie", "age": 40},
]

```python
mistral + 'Please generate a list of 3 jsons with name and age:\n' + simple_list_of_jsons()
```
>Please generate a list of 3 jsons with name and age:
[
{"name": "John", "age": 20},
{"name": "Jane", "age": 25},
{"name": "Bob", "age": 30},
]

**Reply by motaatmo** [reply]
Thank you! Can you say when "one point" is going to be? This would influence my decision whether or not to start coding something that works the (quite simple) json structs that I have to process.
